precompiles/staking/method_claim_rewards.go

```diff
@@ -0,0 +1,147 @@
+	precompiletypes "github.com/zeta-chain/node/precompiles/types"
+	fungibletypes "github.com/zeta-chain/node/x/fungible/types"
+)
+
+// claimRewards claims all the rewards for a delegator from a validator.
+// As F1 Cosmos distribution scheme implements an all or nothing withdrawal, the precompile will
+// withdraw all the rewards for the delegator, filter ZRC20 and unlock them to the delegator EVM address.
+func (c *Contract) claimRewards(
+	ctx sdk.Context,
+	evm *vm.EVM,
+	contract *vm.Contract,
+	method *abi.Method,
+	args []interface{},
+) ([]byte, error) {
+	if len(args) != 2 {
+		return nil, &precompiletypes.ErrInvalidNumberOfArgs{
+			Got:    len(args),
+			Expect: 2,
+		}
+	}
+
+	delegatorAddr, validatorAddr, err := unpackClaimRewardsArgs(args)
+	if err != nil {
+		return nil, err
+	}
+
+	var (
+		// This represents the delegator calling directly the precompile.
+		callerIsDelegator = contract.CallerAddress == delegatorAddr
+
+		// This represents the delegator calling the precompile through a contract.
+		originIsDelegator = evm.Origin == delegatorAddr
+	)
+
+	// If the delegator is not the origin nor the caller, it's an unauthorized operation.
+	if !callerIsDelegator && !originIsDelegator {
+		return nil, precompiletypes.ErrInvalidAddr{
+			Got:    delegatorAddr.String(),
+			Reason: "unauthorized to withdraw the delegation rewards for delegator",
+		}
+	}
+
+	// Get delegator Cosmos address.
+	delegatorCosmosAddr, err := precompiletypes.GetCosmosAddress(c.bankKeeper, delegatorAddr)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get validator Cosmos address.
+	validatorCosmosAddr, err := sdk.ValAddressFromBech32(validatorAddr)
+	if err != nil {
+		return nil, err
+	}
+
+	// Withdraw all the delegation rewards.
+	// The F1 Cosmos distribution scheme implements an all or nothing withdrawal.
+	// The coins could be of multiple denomination, and a mix of ZRC20 and Cosmos coins.
+	coins, err := c.distributionKeeper.WithdrawDelegationRewards(ctx, delegatorCosmosAddr, validatorCosmosAddr)
+	if err != nil {
+		return nil, precompiletypes.ErrUnexpected{
+			When: "WithdrawDelegationRewards",
+			Got:  err.Error(),
+		}
+	}
+
+	// For all the ZRC20 coins withdrawed:
+	// - Check the amount to unlock is valid.
+	// - Burn the Cosmos coins.
+	// - Unlock the ZRC20 coins.
+	for _, coin := range coins {
+		// Filter out invalid coins.
+		if !coin.IsValid() || !coin.Amount.IsPositive() || !precompiletypes.CoinIsZRC20(coin.Denom) {
+			continue
+		}
+
+		// Notice that instead of returning errors we just skip the coin. This is because there might be
+		// more than one ZRC20 coin in the delegation rewards, and we want to unlock as many as possible.
+		// Coins are locked in the bank precompile, so it should be possible to unlock them afterwards.
+		var (
+			zrc20Addr   = common.HexToAddress(strings.TrimPrefix(coin.Denom, config.ZRC20DenomPrefix))
+			zrc20Amount = coin.Amount.BigInt()
+		)
+
+		// Check if bank address has enough ZRC20 balance.
```

Comment:
> // Check if bank address has enough ZRC20 balance. 

This is supposed to always be the case? We could have an invariant in the future but for now we should at least have a log here if it must not happen

Reply:
That's my view on this too. Ideally the locked amount for any existing ZRC20 would be tracked by a `Locked` field in the ZRC20 struct. That `Locked` field can be updated or checked at any given moment from the bank and staking precompile, whenever someone deposits, withdraw or claimRewards.

Reply:
Added in 2b130fdc0f29a4d958d599d633c42f3987a6e7b5